documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+
+The cases mentioned do not apply inside an expression. However, it is in general unsafe to propagate such box outside of an expression evaluation. This means that if `A(10:20:2, :)` if a top level expression, it is probably safer to materialize into a temporary storage and leave later optimizations determine if it is safe to keep skip the materialization.
+This means for the Fortan code:
+```
+B = A(1:10, :)
+PRINT *m, B
+```
+The bridge to FIR will not replace refences to `B` by the `fir.box` encoding ` A(1:10, :)` because it cannot ensure `A` was not modified in-between without an analysis of A potential definition in-between that does not belong to the bridge towards FIR.
+
+An alternative, that requires FIR language modifications, would be to allow `fir.box` to embox array SSA values (`fir.array<T>`) instead of only memory references (like `fir.ref<fir.array<T>>` ).  The value-based aspect would remove such need of performing re-definition analysis to propagate boxes. 
+#### Lowering transformational intrinsic procedures to `fir.box`
+`fir.box` can also be used to implement some transformational intrinsic because it can describe more than array subsection in case it is used with an mlir affine map instead of susbscripts.
+For instance `transpose(A)` where `A` is `REAL, DIMENSION(20, 10)` can be written in FIR as:
+```
+#transpose = (i, j)[] -> (j, i)
+%atrans= fir.embox %a  [#transpose]
+```
+Element `(i, j)` form `transpose(A)` would then simply be accessed with `fir.extract_value %atrans, %i, %j`.
+
+Affine maps can be combined and should allow to implement many of the ideas from the [Array Expression Evaluation Design Document](documentation/ArrayComposition.md) in FIR.
+
+TODO: write the affine maps for all the transformational mentioned in last document.
+
+More can be read about MLIR affine maps in the [Affine Dialect Language Reference](https://github.com/tensorflow/mlir/blob/master/g3doc/Dialects/Affine.md)
+#### Expression storage in FIR
+
+A `fir.loop` is more or less a regular counted loop and it is not returning an ssa value that would be the result of the array expression associated to the loop. Hence, it can only have effects on the program through memory accesses. `fir.store` operations in the `fir.loop` must store the array result elements into some memory reference.
+This implies that every `fir.loop` computing an array expression must be given a memory reference to some storage can store the array result. The required size of such storage may know at compile time or may be dynamic.
+Hence, expression lowering will have to handle storage for expressions and may have to introduce dynamic memory allocation and free operations (`fir.allocmem` and `fir.freemem`).
+To simplify the design, we will assume the entry point of array expression lowering is provided with a memory-reference that can store the result. Whether this reference is a temporary and whether it is dynamic should not affect the lowering.
```

Comment:
I have added sentence: "For instance, when lowering assignment `X = Y + B`, it is the part of the bridge that lowers `parser::AssignStmt` that should be responsible for the storage required to store `Y + B`. This is to de-correlate as much as possible expression lowering from the statements in which expressions may appear."


To answer you question: in some ways, yes, there is something before expression lowering that would do this. However it is not a pass done from the expression lowering facility that would that, but the lowering of the statement where the expression appears (it may require doing some kind of pass over the expression to get some information like the shape with `evaluate::GetSahpe`).

If  `parser::AssignStmt` lowering (that is not done by expression lowering)  allocates a dynamic temporary for the right-hand side evaluation, then it should also be the one the one inserting the clean-up after the copy to the left hand side variable is done. But this should be transparent to expression lowering.

I have some other ideas to do this de-correlation. It would involve the expression facility returning or accepting lambdas that would apply the array expressions given the array indexes passed as argument to the lambda. But that is still in aplha version in my head.